litr-filters/src/main/java/com/linkedin/android/litr/filter/video/gl/AnimatedGifOverlayFilter.java

```diff
@@ -36,64 +21,55 @@
 
     private static final String TAG = AnimatedGifOverlayFilter.class.getSimpleName();
 
+    private final AnimationFrameProvider animationFrameProvider;
+
     private Frame currentFrame;
     private long nextFramePresentationTime;
 
     /**
      * Create filter with certain configuration.
-     * @param context context for accessing bitmap
-     * @param bitmapUri bitmap {@link Uri}
+     * @param animationFrameProvider {@link AnimationFrameProvider} which provides animation frames and their durations
      * @param bitmapRect Rectangle of bitmap's target position on a video frame, in relative coordinate in 0 - 1 range
      *                   in fourth quadrant (0,0 is top left corner)
      */
-    public AnimatedGifOverlayFilter(@NonNull Context context, @NonNull Uri bitmapUri, @Nullable RectF bitmapRect) {
-        super(context, bitmapUri, bitmapRect);
+    public AnimatedGifOverlayFilter(@NonNull AnimationFrameProvider animationFrameProvider, @Nullable RectF bitmapRect) {
+        super(bitmapRect);
+
+        this.animationFrameProvider = animationFrameProvider;
     }
 
     @Override
     public void init(@NonNull float[] mvpMatrix, int mvpMatrixOffset) {
         super.init(mvpMatrix, mvpMatrixOffset);
 
-        try {
-            ContentResolver contentResolver = context.getContentResolver();
-            InputStream inputStream = contentResolver.openInputStream(bitmapUri);
-            BitmapPool bitmapPool = new LruBitmapPool(10);
-            GifBitmapProvider gifBitmapProvider = new GifBitmapProvider(bitmapPool);
-            GifDecoder gifDecoder = new StandardGifDecoder(gifBitmapProvider);
-            gifDecoder.read(inputStream, (int) getGifSize(bitmapUri));
-
-            int frameCount = gifDecoder.getFrameCount();
-            Frame firstFrame = null;
-            Frame prevFrame = null;
-            Frame frame = null;
-            Bitmap frameBitmap;
-            for (int frameIdx = 0; frameIdx < frameCount; frameIdx++) {
-                gifDecoder.advance();
-                frameBitmap = gifDecoder.getNextFrame();
-                if (frameBitmap == null) {
-                    Log.e(TAG, "Error loading GIF frame " + frameIdx);
-                    continue;
-                }
-                int textureId = createOverlayTexture(frameBitmap);
-                frame = new Frame(textureId, gifDecoder.getNextDelay());
-                if (frameIdx == 0) {
-                    firstFrame = frame;
-                }
-                if (prevFrame != null) {
-                    prevFrame.next = frame;
-                }
-                prevFrame = frame;
-                frameBitmap.recycle();
+        Frame firstFrame = null;
+        Frame prevFrame = null;
+        Frame frame = null;
+        Bitmap frameBitmap;
+        for (int frameIdx = 0; frameIdx < animationFrameProvider.getFrameCount(); frameIdx++) {
```

Comment:
A list of bitmaps may or may not cause performance issue, for the same reason that a gif decoder won't decode all bitmaps in one shot, instead 2-3 when needed. Can we do stress test and measure the performance, so no more concern?

Reply:
Each bitmap gets recycled once they are converted into a GL texture to free up memory.

Reply:
That's correct. But this design is assuming that all bitmaps need to be decoded then passed in. Think about a common case, showing gif, it getNextFrame/decodes one frame, show it, the recycle it, which has no concern. If you decode all then pass them, it might be a problem.